src/pybind/mgr/mgr_util.py

```diff
@@ -335,6 +338,116 @@ def get_all_filesystems(self) -> List[str]:
         return fs_list
 
 
+class CephFsEarmarkResolver:
+    """
+    Resolves and manages earmarks for CephFS subvolumes.
+
+    This class provides functionality to fetch and manage earmarks for subvolumes,
+    ensuring consistent handling across different modules (NFS, SMB).
+    """
+    def __init__(self, mgr: Module_T, *, client: Optional[CephfsClient] = None) -> None:
+        self._mgr = mgr
+        self._cephfs_client = client or CephfsClient(mgr)
+
+    def _fetch_data(self, command: Dict[str, str], ctx: str) -> Optional[List[Dict[str, str]]]:
+        """Fetch data using the provided command and return a list of dictionaries."""
+        logger.debug('Fetching %s: %r', ctx, command)
+        ret, data, status = self._mgr.mon_command(command)
+        if ret != 0:
+            logger.error('Failed to fetch %s: %r', ctx, status)
+            return None
+        logger.info('Fetched %s: %r', ctx, data)
+        try:
+            return json.loads(data.strip())
+        except json.JSONDecodeError:
+            logger.error('Failed to parse JSON for %s: %r', ctx, data)
+            return None
+
+    def _fetch_subvolume_groups(self, volume: str) -> List[str]:
+        """Fetch subvolumegroups for a given volume."""
+        cmd = {'prefix': 'fs subvolumegroup ls', 'vol_name': volume}
+        data = self._fetch_data(cmd, "subvolumegroups")
+        return [subvolgrp["name"] for subvolgrp in data] if data else []
+
+    def _fetch_subvolumes(self, volume: str, subvolumegrp: str) -> List[str]:
+        """Fetch subvolumes for a given volume and subvolumegroup."""
+        cmd = {
+            'prefix': 'fs subvolume ls',
+            'vol_name': volume,
+            'group_name': subvolumegrp if subvolumegrp != '_nogroup' else '',
+        }
+        data = self._fetch_data(cmd, "subvolumes")
+        return [subvol["name"] for subvol in data] if data else []
+
+    def _fetch_from_path(self, path: str, names: List[str], default: Optional[str] = None) -> Optional[str]:
+        """Extract a name from a path using the provided list of names."""
+        path_parts = path.split('/')
+        for part in path_parts:
+            if part in names or (default and part == default):
+                return part
+        return None
+
+    def _fetch_subvolumegroup_from_path(self, volume: str, path: str) -> Optional[str]:
+        """Extract subvolumegroup from a path."""
+        subvolumegroups = self._fetch_subvolume_groups(volume)
+        return self._fetch_from_path(path, subvolumegroups, '_nogroup')
+
+    def _fetch_subvolume_from_path(self, volume: str, path: str, subvolumegroup: Optional[str] = None) -> Optional[str]:
+        """Extract subvolume from a path."""
+        subvolumes = self._fetch_subvolumes(volume, subvolumegroup or '')
+        return self._fetch_from_path(path, subvolumes)
```

Comment:
Again, this feels like the resolver stuff that already exists.

Reply:
AFAIK we're resolving path right? Here we need to fetch subvol/grp name from given path which we don't hv anywhere implemented yet

Reply:
Fair enough. But we should not need this feature for smb then. As the smb module currently only needs https://github.com/ceph/ceph/blob/15cbae0a2745e79e6a39d694b39e5140428c3148/src/pybind/mgr/smb/fs.py#L64 to map *to* a path not from a path.

We need to minimize round trips between the modules to achieve the performance needed when we have to validate many shares.

Reply:
yes, I fixed that in recent commit